lib/src/clojure_lsp/feature/call_hierarchy.clj

```diff
@@ -11,14 +11,25 @@
 
 (set! *warn-on-reflection* true)
 
+(defn ^:private safe-zloc-of-forced-uri [db* uri]
+  (some-> (f.file-management/force-get-document-text uri db*)
+          (parser/safe-zloc-of-string)))
+
+;; TODO: call-hierarchies could be made much faster if
+;; element-by-uri->call-hierarchy-item didn't re-parse so often.
+;; 1. Reason through whether it really needs the branch that parses. The tests
+;;    suggest it doesn't.
+;; 2. Failing that, when dep-graph is merged, use it to find ns name.
+;; 3. Failing that, gather URIs that need to be parsed up front, parse each one
+;;    once, then thread those root zlocs through the rest of the code.
 (defn ^:private element-by-uri->call-hierarchy-item
-  [{uri :uri
-    {:keys [ns filename name-row name-col arglist-strs deprecated] el-name :name :as parent-element} :parent-element
-    usage-element :usage-element} db*]
-  (let [project-file? (string/starts-with? uri "file://")
+  [{:keys [uri parent-element usage-element]} db*]
+  (let [{:keys [ns filename name-row name-col arglist-strs deprecated] el-name :name} parent-element
+        project-file? (string/starts-with? uri "file://")
         detail (if project-file?
-                 (some-> (f.file-management/force-get-document-text uri db*)
-                         (parser/zloc-of-string) ;; throws on invalid Clojure
+                 ;; TODO: tests pass if this branch is removed. Why can't we use
+                 ;; the else branch for project files?
```

Comment:
Good question, I can't recall why I added that at the time, but if we can always get the namespace reliable and fallback for filenames when not seems already good enough, so the else branch makes sense.
Also, parsing the loc just to find the namespace name sounds too much

Reply:
Looks like the parsing was added all the way back in bc39d0ed7df803bde86dbcc7638e9ef6754f5484. It was introduced before we started using `ns`, which was added in 2e154faccf38652351bfeff1717a719818f9397a. I think everything could have been switched to `ns` at that time. I'll make the switch now, keeping the filename fallback. 🎉

Reply:
That makes sense! Nice